reports/glides_validator.livemd

```diff
@@ -145,10 +213,426 @@ table_ui =
+# Do we need to worry about cases where actual departure happened around the start/end of the hour?
+# Maybe give a little wiggle room into the next/previous hour, for predictions..?
+
+predicted_departures_by_hour =
+  trip_updates
+  # Group by hour of the predicted departure (not hour the prediction was generated!),
+  # in local time.
+  # This requires splitting each trip update into its individual stop_time_update items.
+  |> Stream.flat_map(fn tr_upd ->
+    Enum.map(
+      tr_upd.trip_update.stop_time_update,
+      &%{trip_id: tr_upd.id, departure_time: &1.departure.time}
+    )
+  end)
+  |> Enum.group_by(
+    &unix_timestamp_to_local_hour.(&1.departure_time),
+    & &1.trip_id
+  )
+  |> Map.new(fn {hour, trip_ids} -> {hour, MapSet.new(trip_ids)} end)
+
+# |> IO.inspect(label: "predicted")
+
+actual_departures_by_hour =
+  vehicle_positions
+  # Group by hour
+  |> Enum.group_by(
+    &unix_timestamp_to_local_hour.(&1.vehicle.timestamp),
+    & &1.vehicle.trip.trip_id
+  )
+  |> Map.new(fn {hour, trip_ids} -> {hour, MapSet.new(trip_ids)} end)
+
+# |> IO.inspect(label: "actual")
+
+0..23
+|> Enum.map(fn hour ->
+  predicted_departure_trips = Map.get(predicted_departures_by_hour, hour, MapSet.new())
+  actual_departure_trips = Map.get(actual_departures_by_hour, hour, MapSet.new())
+
+  %{
+    "hour" => "#{pad.(hour)}:00",
+    "# of predicted departure trips" => MapSet.size(predicted_departure_trips),
+    "# of actual departure trips" => MapSet.size(actual_departure_trips)
+  }
+end)
+|> Kino.DataTable.new(
+  keys: ["hour", "# of predicted departure trips", "# of actual departure trips"]
+)
+
+# TODO: Rotate results so they start at beginning of the service day, rather than midnight?
+```
+
+## Per-Hour % of trips for which RTR made departure predictions vs. actual departures
+
+```elixir
+0..23
+|> Enum.map(fn hour ->
+  predicted_departures = Map.get(predicted_departures_by_hour, hour, MapSet.new())
+  actual_departures = Map.get(actual_departures_by_hour, hour, MapSet.new())
+
+  # *** For reviewer: ***
+  # This does trip matching, which the ticket description said is not needed.
+  # I'm not sure how else to interpret the requested metric, though--am I
+  # misunderstanding it, or is the ticket description incorrect?
+  actual_departures_with_predictions =
+    MapSet.intersection(predicted_departures, actual_departures)
+
+  actual_departure_count = MapSet.size(actual_departures)
+  actual_departure_with_prediction_count = MapSet.size(actual_departures_with_predictions)
+
+  percentage =
+    if actual_departure_count > 0 do
+      (100.0 * (actual_departure_with_prediction_count / actual_departure_count))
+      |> round()
+      |> then(&"#{&1}%")
+    else
+      "N/A (0 actual departures)"
+    end
+
+  %{
+    "hour" => "#{pad.(hour)}:00",
+    "% of actual departures that had predictions" => percentage
+  }
+end)
+|> Kino.DataTable.new(keys: ["hour", "% of actual departure trips that had predictions"])
 ```
 
```

Comment:
I've tried a couple of runs on different dates and I don't seem to have the percentage column on this one.

![Screenshot 2024-06-13 at 2 47 23 PM](https://github.com/mbta/transit_data_reports/assets/304505/87049108-83a1-47dc-b1ee-c3111d775fae)

Reply:
Oh, very strange. I'm not sure how the current code could fail to produce a non-empty string for that column. Do you know what date / configuration resulted in this?

Reply:
Oops. I must have renamed the map key at some point without also updating it in the `DataTable.new` call.

Fixed now!